File: frontend/src/types/project.ts
/**
 * プロジェクトの共同作業者情報を定義するインターフェース
 */
export interface Collaborator {
  id: number;           // 共同作業レコードのID
  user_id: number;      // ユーザーID
  user_email?: string;  // ユーザーのメールアドレス
  permission: 'viewer' | 'editor'; // 権限 (viewer: 閲覧のみ, editor: 編集可能)
}

/**
 * プロジェクトの情報を定義するインターフェース
 */
export interface Project {
  id: number;           // プロジェクトID
  name: string;         // プロジェクト名
  description?: string; // プロジェクトの説明 (任意)
  created_at: string;   // 作成日時
  updated_at: string;   // 最終更新日時
  owner_id: number;     // プロジェクト所有者のユーザーID
  collaborators?: Collaborator[]; // 共同作業者のリスト
}

/**
 * ダッシュボードなどで表示するためのプロジェクト概要
 * Projectインターフェースを拡張し、統計情報を追加しています。
 */
export interface ProjectSummary extends Project {
  todo_count: number;      // プロジェクト内の総タスク数
  completed_count: number; // 完了済みのタスク数
  role?: 'owner' | 'collaborator'; // 現在のユーザーのプロジェクトにおける役割
}

/**
 * プロジェクトを新規作成する際に必要なデータ
 */
export interface CreateProjectData {
  name: string;         // プロジェクト名 (必須)
  description?: string; // 説明 (任意)
}

/**
 * プロジェクト情報を更新する際に使うデータ
 * CreateProjectDataの全項目をオプションにしたものです。
 */
export type UpdateProjectData = Partial<CreateProjectData>;

File: backend/app/schemas/project.py
from __future__ import annotations
from datetime import datetime
from pydantic import BaseModel, ConfigDict

class ProjectBase(BaseModel):
    """
    プロジェクトの基本情報を定義するスキーマ。
    """
    name: str                       # プロジェクト名 (必須)
    description: str | None = None  # プロジェクトの説明 (任意)

class ProjectCreate(ProjectBase):
    """
    プロジェクト新規作成用のスキーマ。
    """
    pass

class ProjectUpdate(BaseModel):
    """
    プロジェクト情報更新用のスキーマ。全項目が任意。
    """
    name: str | None = None         # プロジェクト名
    description: str | None = None  # 説明

class CollaboratorBase(BaseModel):
    """
    共同作業者の基本情報を定義するスキーマ。
    """
    user_id: int                    # ユーザーID
    permission: str = "editor"      # 権限 (デフォルト: "editor")

class CollaboratorCreate(CollaboratorBase):
    """
    共同作業者追加用のスキーマ。
    """
    pass

class CollaboratorOut(CollaboratorBase):
    """
    APIから返却される共同作業者のデータ構造。
    """
    id: int                         # インデックスID
    user_email: str | None = None   # フロントエンド表示用のメールアドレス
    
    model_config = ConfigDict(from_attributes=True)

class ProjectOut(ProjectBase):
    """
    APIから返却されるプロジェクト情報のフルセット。
    """
    id: int                         # プロジェクトID
    created_at: datetime            # 作成日時
    updated_at: datetime            # 更新日時
    owner_id: int                   # オーナーのユーザーID
    organization_id: int | None = None # 所属組織ID
    collaborators: list[CollaboratorOut] = [] # 共同作業者のリスト
    
    model_config = ConfigDict(from_attributes=True)

class ProjectSummary(ProjectOut):
    """
    ダッシュボード用などのプロジェクトサマリー
    """
    todo_count: int = 0
    completed_count: int = 0
    role: str | None = None
